File: java/PH.java
package irresponsible.crouton;

import clojure.lang.IPersistentVector;
import clojure.lang.ITransientMap;
import clojure.lang.RT;

public abstract class PH extends AInvariant {
  protected final Object name;
  protected PH(Object n, IRoute nxt) {
    super(nxt);
    if (null == n)
      throw new IllegalArgumentException("name must not be nil");
    name = n;
  }
  final protected Object forward(Object val, IPersistentVector pieces, ITransientMap places) {
    return next.match(RT.subvec(pieces, 1, pieces.count()),
		      places.assoc(name, val));
  }
};

File: kotlin/src/PH.kt
package irresponsible.crouton

import clojure.lang.IPersistentVector
import clojure.lang.IPersistentMap
import clojure.lang.RT

abstract class PH protected constructor(protected val name: Any, nxt: IRoute) : AInvariant(nxt) {
    protected fun forward(value: Any, pieces: IPersistentVector, places: IPersistentMap): Any? {
        return next.match(RT.subvec(pieces, 1, pieces.count()),
                places.assoc(name, value))
    }
}
